### PR TITLE
Abort transfer on force peer delete

### PR DESCRIPTION
**Pull request by KShivendu**
Force deletes to a peer should also abort all shards transfers from this peer. Otherwise, they can remain stuck there forever.

TODO:
- [ ] Think we should abort `to` transfers for this force deleted peer

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### CODE REVIEW

File: lib/storage/src/content_manager/toc/mod.rs [hunk 1]
```diff
@@ -461,6 +461,26 @@ impl TableOfContent {
         false
     }
 
+    pub async fn abort_peer_transfers(&self, peer_id: PeerId) -> CollectionResult<()> {
+        for collection in self.collections.read().await.values() {
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply lib/storage/src/content_manager/toc/mod.rs:465]: > Because removing it from the state here, does not stop the ongoing transfer task

Ahh right. I didn't realise that `register_abort_transfer()` is local and it's not pushing it to consensus. Thanks :)

[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/storage/src/content_manager/toc/mod.rs:473]: `remove_peer` is a consensus operation itself, and so it will be called on all peers through consensus, so we don't need to `submit_collection_meta_op`, we can just remove transfer related data/stop transfer tasks *locally*.

Arguably, `submit_collection_meta_op` here is wrong, because it will (I think?) cause *every* peer to submit a copy of `ShardTransferOperations::Abort`.

File: src/actix/api/cluster_api.rs [hunk 2]
```diff
@@ -79,6 +79,13 @@ fn remove_peer(
             });
         }
 
+        // If we reach here, either there are no shards or force removal is requested
+        if has_shards {
+            log::warn!("Aborting peer transfers for peer {peer_id} before force removal");
+            toc.abort_peer_transfers(dispatcher.clone(), access, peer_id)
+                .await?;
+        }
```
Comment (ffuugoo) [inline src/actix/api/cluster_api.rs:87]: IMO, this should be done differently. Instead of explicitly submitting `ShardTransferOperations::Abort` in the API, we should modify the code that is called when `RemovePeer` consensus operation is applied, and make it abort transfers as part of `RemovePeer` operation.

File: lib/collection/src/collection/mod.rs [hunk 3]
```diff
@@ -580,6 +580,10 @@ impl Collection {
             }
         }
 
+        for transfer in self.get_related_transfers(peer_id).await {
+            self.abort_shard_transfer(transfer.key(), None).await?;
```
Comment (KShivendu) [inline lib/collection/src/collection/mod.rs:584]: Edge case: What if you read about a transfer and it finished just before you try to delete it. Has happened in previous test failure. https://github.com/qdrant/qdrant/actions/runs/14967895398/job/42041945647?pr=6507

**Review by timvisee — APPROVED** [review]
I like this a lot better! Simple and effective.

File: lib/collection/src/collection/mod.rs (hunk 3 above)
Comment (timvisee) [reply lib/collection/src/collection/mod.rs:584]: It's fine, it'll be ignored deeper down: https://github.com/qdrant/qdrant/blob/9c723aa48083a4bc5e2d9124b827405b2c35b398/lib/collection/src/collection/shard_transfer.rs#L331-L333
